src/ILCompiler.Compiler/src/Compiler/CompilationBuilder.Aot.cs

```diff
@@ -0,0 +1,71 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace ILCompiler
+{
+    partial class CompilationBuilder
+    {
+        // These need to provide reasonable defaults so that the user can optionally skip
+        // calling the Use/Configure methods and still get something reasonable back.
+        protected MetadataManager _metadataManager;
+        protected InteropStubManager _interopStubManager = new EmptyInteropStubManager();
+        protected VTableSliceProvider _vtableSliceProvider = new LazyVTableSliceProvider();
+        protected DictionaryLayoutProvider _dictionaryLayoutProvider = new LazyDictionaryLayoutProvider();
+        protected DebugInformationProvider _debugInformationProvider = new DebugInformationProvider();
+        protected DevirtualizationManager _devirtualizationManager = new DevirtualizationManager();
```

Comment:
AFAIK CPAOT currently uses a different devirtualization manager than AOT. Are you planning to make it partial too (DevirtualizationManager.Aot / ReadyToRun) or somehow create a unified manager usable by both? Not mentioning that I suspect that the _compilationModuleGroup.ContainsXXX checks in the R2R DevirtualizationManager are actually imprecise and should instead call VersionsWithXXX. [Just to clarify, that is certainly my fault, not yours.]

Reply:
OK, I think I found the answer - the base class initializes the field with the default (common) DevirtMgr and then rewrites it with the R2R version in R2RCompilationBuilder. I guess it's probably fine for the purpose of the port even though the split will stop actually making sense after the port.

Reply:
The file this comment is on is specific to AOT - we don't `new` the base `DevirtualizationManager` in ReadyToRun mode.  In ReadyToRun mode, devirtualization manager is not configurable (seems like all the configurability we need for R2R is already wrapped by devirtualization manager using compilation module group to drive decisions).

The base devirtualization manager could have been abstract, but it doesn't feel like it's hurting anything to just have it do devirtualization based on existing metadata. Besides, most of the children (including the ready to run version of the devirtualization manager) call into `base`, so making it abstract would be inconvenient there.